rustc: Prepare to enable ThinLTO by default

alexcrichton opened the pull request:
This commit *almost* enables ThinLTO and multiple codegen units in release mode by
default but is blocked on #46346 now before pulling the trigger.

rust-highfive commented:
r? @arielb1

(rust_highfive has picked a reviewer for you, use r? to override)

alexcrichton commented:
r? @michaelwoerister 

cc @rust-lang/compiler

sfackler commented:
Is this a blocker? https://github.com/rust-lang/rust/issues/46346

alexcrichton commented:
Are we sure that's related to ThinLTO? I've tested everything locally and wasn't able to reproduce myself.

alexcrichton commented:
Urgh it may very well be. I swear to god.

kennytm commented on src/rustllvm/PassWrapper.cpp:
```diff
@@ -1062,11 +982,27 @@ LLVMRustCreateThinLTOData(LLVMRustThinLTOModule *modules,
+  // summaries in the index, and we basically just only want to ensure that dead
+  // symbols are internalized. Otherwise everything that's already external
+  // linkage will stay as external, and internal will stay as internal.
+  std::set<GlobalValue::GUID> ExportedGUIDs;
```
Why not use a `std::unordered_set` here.

cuviper commented:
I see that `LLVMRustThinLTOAvailable` gates on just `LLVM_VERSION_GE(4, 0)`.  How well are you expecting this to work when using an _external_ LLVM-4.0?
(IOW, are there rust-lang/llvm patches that I definitely need here?)

alexcrichton commented:
@cuviper IIRC https://github.com/rust-lang/llvm/commit/e45c75de1148456a9eb1a67c14a66df4dfb50c94 is the main (and possibly only) ThinLTO related patch (although it fixes a preexisting bug). 

In general though I lose track over time what patch was for what on our fork, although everything that matters is upstream in some form so it's just a matter of time until we update to get it.

cuviper commented:
Ok, thanks, we already grabbed that one. I'll look back over the branch again later myself. I do appreciate that everything is fixed upstream first!

michaelwoerister commented:
Awesome! Will review shortly. Am I reading this right, this is still blocked on #46346?

alexcrichton commented:
@michaelwoerister unfortunately yes :(

michaelwoerister commented on src/librustc/session/mod.rs:
```diff
@@ -656,30 +656,89 @@ impl Session {
+    }
+
+    /// Returns whether ThinLTO is enabled for this compilation
+    /// compilation
```
compilation compilation

michaelwoerister commented on src/libstd/sys_common/backtrace.rs:
```diff
@@ -252,8 +252,26 @@ fn output_fileline(w: &mut Write,
-    // First validate the symbol. If it doesn't look like anything we're
+pub fn demangle(writer: &mut Write, mut s: &str, format: PrintFormat) -> io::Result<()> {
+    // During ThinLTO LLVM may import and rename internal symbols, so strip out
+    // those endings first as they're on of the last manglings applied to symbol
```
on -> one

michaelwoerister commented on src/rustllvm/PassWrapper.cpp:
```diff
@@ -1004,12 +925,10 @@ LLVMRustCreateThinLTOData(LLVMRustThinLTOModule *modules,
   for (int i = 0; i < num_symbols; i++) {
-    addPreservedGUID(Ret->Index,
-                     Ret->GUIDPreservedSymbols,
-                     GlobalValue::getGUID(preserved_symbols[i]));
```
Was this too conservative before?

alexcrichton commented:
@michaelwoerister ok I've updated the listing of commits and addressed comments. The commits now to *not* enable ThinLTO by default but instead retain retain today's default. The repo is left in a state though to make it a one line change to turn ThinLTO on by default in the future.

To answer your question about `GUIDPreservedSymbols` though there's more info on https://github.com/rust-lang/rust/issues/46374 but it boils down to two things:

* In the first location where `GUIDPreservedSymbols` was used the logic of `addPreservedGUID` wasn't necessary. The same result is computed whether we call `addPreservedGUID` or not.
* In the second location it was *incorrect* to call `addPreservedGUID` in the sense that it added the transitive closure of symbols to that set. That in turn caused https://github.com/rust-lang/rust/issues/46374 where too many symbols were exported (aka all of them) rather than just the set we wanted. By avoiding `addPreservedGUID` and tweaking the definition of `isExported` this bug was fixed though.

michaelwoerister commented:
Thanks for the explanation!

@bors r+

bors commented:
:pushpin: Commit 855f6d1 has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister 855f6d1483e023cea3b7988db294ed9767e15359 -->

bors commented:
:hourglass: Testing commit 855f6d1483e023cea3b7988db294ed9767e15359 with merge 088f328ec5839d535891de5a5af55a6ba3479198...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5581), [status-travis](https://travis-ci.org/rust-lang/rust/builds/310642618?utm_source=github_status&utm_medium=notification)
Approved by: michaelwoerister
Pushing 088f328ec5839d535891de5a5af55a6ba3479198 to master...
